ポリモーフィック関連付けを使用したコメント機能の追加

yama998877 opened the pull request:
一覧、個別、作成、編集、削除のできる日報登録機能を作成後、コメント機能を作成して、本と日報の個別ページにコメントフォームとその個別ページに投稿されたコメントが一覧で表示されるようにしました。

yama998877 committed:
日報のCRUDを作成

コマンドで作成したテンプレート
jbuilderとテストファイルは含まない。

yama998877 committed:
Reportsテーブルのカラムのデータ型を間違えていたので修正

`body`のデータ型を`string`から`text`に修正

yama998877 committed:
辞書ファイルに日報機能で使用する訳文を追加

yama998877 committed:
時刻の表示を日本時間に変更

yama998877 committed:
日報のCRUDを要件を満たすように変更

日報を作成したユーザーでないと編集ページへの移動、更新、削除をできないように変更。
ページング処理を追加。
日報作成時にログインしているユーザーが日報の作成者になるように変更。
ユーザー削除時はそのユーザーが作成した日報が削除されるように変更
ビューのメニュー欄に日報を追加
現在ログインしているユーザーと日報を作成したユーザーが一致しない場合、編集、削除のリンクが表示されないように変更
ユーザー名がない場合はメールアドレスが日報に表示されるように変更
ビューの国際化するように変更
CSSを適用

yama998877 committed:
コメント機能のデータテーブルを作成

yama998877 committed:
ポリモーフィック関連付けを用いたコメント機能

comments_controller.rbにコメント作成と削除のメソッドを追加。削除はコメントを作成したユーザーとログインしているユーザーが一致するとできるように条件を追加。
ユーザー名がなければユーザーメールアドレスが表示されるヘルパーメソッドをusers_helper.rbに追加。
bookモデルがコメント機能を使えるようにbook.rbにコードを追加。
コマンド生成されたcomment.rbにバリデーションを追加。
reportモデルがコメント機能を使えるようにコード追加。
`books/show.html.erb`と`reports/show.html.erb`にコメントリスト(`_list.html.erb`)とコメントフォーム(`_form.html.erb`)を追加。
コメント機能を使えるようにルーティングをネスト。
辞書ファイルにキーと訳文を追加。
ユーザーデータ削除時にそのユーザーがしたコメントが削除されるように設定。

yama998877 committed:
日報機能の軽微な修正

reports_controller.rb内の不要なコメントの削除。
_report.html.erbの日報の投稿日時の表示に`localize`メソッドを使用するように修正。
日報の作成者を表示するコードを作成したヘルパーメソッドを使用するように修正。

yuuu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,53 @@
+    @comment = Comment.find(params[:id])
+  end
+
+  def user_confirmation
```
メソッド名は動詞から始まるようにしたいですね 🙏

yama998877 replied:
ご指摘ありがとうございます🙏
動詞から始まるように修正しました。

yuuu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,53 @@
+    @comment.user_id = current_user.id
+    if @comment.save
+      redirect_to polymorphic_path(@commentable), notice: t('controllers.common.notice_create', name: Comment.model_name.human)
+    elsif @comment.commentable_type == 'Report'
```
このコントローラには複数箇所で「本か、日報か」という判定が存在しています。
今後、コメント対象が本・日報以外に増えると、あっという間に肥大化する懸念があるので、対策を打ちたいです。

対策案として、このコントローラを継承して「本のコメント」「日報のコメント」を処理するコントローラをそれぞれ作成し、ルーティングもそのコントローラに向けてはどうでしょう？
その際、共通する処理は親クラスに残したままとしたいです 🙏

yama998877 replied:
ご指摘ありがとうございます🙏
継承したコントローラを作成し、ルーティングも修正してコントローラ内の条件分岐が解消できました。

yuuu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,53 @@
+
+  # DELETE /comments/1
+  def destroy
+    @comment.destroy
```
ここは、DBの障害等で万が一のエラーが起きた場合に気付けるよう `destroy!` を用いた方が良いです。
日報の削除についても同様です。

yama998877 replied:
ご指摘ありがとうございます。
`destroy!`を使用するように修正しました🙇‍♂️

yuuu commented on app/helpers/comments_helper.rb:
```diff
@@ -0,0 +1,4 @@
+# frozen_string_literal: true
+
+module CommentsHelper
```
使う予定のないHelperは削除しておきましょう 🙏

yama998877 replied:
ご指摘ありがとうございます🙏
削除しました🙇‍♂️

yuuu commented on app/helpers/users_helper.rb:
```diff
@@ -4,4 +4,12 @@ module UsersHelper
     [user.name, user.email].find(&:present?)
   end
+
+  def name_or_email(resouce)
```
このメソッドを用意するという発想自体は良いですね 👍 
今回はViewでの利用なのでヘルパーでも不自然ではないですが、View以外でも使われうるメソッドかなとも思うので、 `User` モデルに定義しても良さそうです。

yama998877 replied:
ご指摘ありがとうございます。
`User`モデルに定義し直して、全角スペースのみのユーザー名に対応できてなかったのでコードを修正しました。

yuuu commented on app/views/comments/_form.html.erb:
```diff
@@ -0,0 +1,18 @@
+<%= form_with model: [commentable, comment], local: true do |f| %>
+  <% if comment.errors.any? %>
+    <div style="color: red">
+      <h2><%= t('views.common.validation_error', errors: i18n_error_count(comment.errors.count), name: Comment.model_name.human.downcase) %></h2>
```
エラーメッセージの表示をそれぞれのViewに記載しているので、partialやhelperを用いて共通化すると良いですね。

yama998877 replied:
ご指摘ありがとうございます。
複数のViewで使用するので、新たに`shared`ディレクトリを作成し、エラーメッセージの partialファイルをそのディレクトリに作成して共通化しました。

yama998877 committed:
修正

`set_commentable`を`CommentsController`から削除して条件分岐をせずに済むように`Books::CommentsController`と`Reports::CommentsController`に移動。
`create`メソッドの`elseif`を削除し、`else`に変更。`else`内に`render_commentable_show`メソッドを使用。
`render_commentable_show`は`Books::CommentsController`、`Reports::CommentsController`に定義。
`destroy`メソッドを削除失敗時に例外が発生するように`destroy!`に変更。
`user_confirmation`を動詞から始まるようにメソッド名を変更。
使用していないビューヘルパーを削除。
`users_helper.rb`に作成した`name_or_email`メソッドを`user.rb`に移動。
移動した`name_or_email`が全角スペースだけのユーザー名には対応できてなかったので修正。
各ビューファイルにエラーメッセージのコードが散らばっていたのを`views/shared/form_error_messages.html.erb`に統一して`render`で呼び出せるように修正。
books,reportsのshowページが表示されるように部分テンプレート名を変更。
モデルファイルに移動させた`name_or_email`メソッドをビューで使えるように修正。
`books`と`reports`のルーティングを名前空間付きコントローラー`Books::CommentsController`,`Reports::CommentsController`を使うように修正。
